lib/spack/llnl/util/filesystem.py

```diff
@@ -265,25 +267,112 @@ def unset_executable_mode(path):
-    """Manually install a directory tree to a particular location."""
-    tty.debug("Installing %s to %s" % (src, dest))
-    shutil.copytree(src, dest, **kwargs)
+def copy_tree(src, dest, symlinks=True):
+    """Recursively copy an entire directory tree rooted at *src*.
+
+    If the destination directory *dest* does not already exist, it will
+    be created as well as missing parent directories.
+
+    If *symlinks* is true, symbolic links in the source tree are represented
+    as symbolic links in the new tree and the metadata of the original links
+    will be copied as far as the platform allows; if false, the contents and
+    metadata of the linked files are copied to the new tree.
+
+    Parameters:
+        src (str): the directory to copy
+        dest (str): the destination directory
+        symlinks (bool): whether or not to preserve symlinks
+    """
+    tty.debug('Copying {0} to {1}'.format(src, dest))
+
+    mkdirp(dest)
+
+    for s, d in traverse_tree(src, dest, order='pre', follow_nonexisting=True):
+        if symlinks and os.path.islink(s):
+            # Note that this won't rewrite absolute links into the old
+            # root to point at the new root. Should we handle that case?
+            target = os.readlink(s)
+            os.symlink(os.path.abspath(target), d)
+        elif os.path.isdir(s):
+            mkdirp(d)
+        else:
+            shutil.copyfile(s, d)
+
+
+def install_tree(src, dest, symlinks=True):
+    """Recursively install an entire directory tree rooted at *src*.
+
+    If the destination directory *dest* does not already exist, it will
+    be created as well as missing parent directories.
+
+    If *symlinks* is true, symbolic links in the source tree are represented
+    as symbolic links in the new tree and the metadata of the original links
+    will be copied as far as the platform allows; if false, the contents and
+    metadata of the linked files are installed to the new tree.
+
+    Works similarly to :py:func:`copy_tree` with the addition of setting proper
+    permissions on the installed files and directories.
+
+    Parameters:
+        src (str): the directory to install
+        dest (str): the destination directory
+        symlinks (bool): whether or not to preserve symlinks
+    """
+    tty.debug('Installing {0} to {1}'.format(src, dest))
+
+    mkdirp(dest)
+
+    for s, d in traverse_tree(src, dest, order='pre', follow_nonexisting=True):
+        if symlinks and os.path.islink(s):
+            # Note that this won't rewrite absolute links into the old
+            # root to point at the new root. Should we handle that case?
+            target = os.readlink(s)
+            os.symlink(os.path.abspath(target), d)
+        elif os.path.isdir(s):
+            mkdirp(d)
+        else:
+            shutil.copyfile(s, d)
 
-    for s, d in traverse_tree(src, dest, follow_nonexisting=False):
         set_install_permissions(d)
         copy_mode(s, d)
```

Comment:
These two lines contradict each other and make it kind of unclear how permissions will be set.  The first sets some arbitrary stuff we decided on in `set_install_permissions`, the second just copies whatever the build system did, and overwrites the first.  So we should do one or the other.

I think the right thing at the moment is to use `copy_mode`, as `set_install_permissions` is almost certainly not the right thing for everything in a tree.  We should probably just get rid of the function.  All it does is set to `755` or `644` depending on whether the file is a directory or a file.  It'll clobber executable bits on executables and other files that need them.  It looks like the only place it is used in `gcc`, for installing a file.

@becker33 is working on some permissions schemes for Spack... I think we should get to a point where we have config options to decide whether the install tree is single-user, group readable, world readable, or group-writable, and what the group should be (if not the owner's default group). Those are the only 4 sensible configurations I can think of.  If we specify that in the config file, I think the right things to do after install will be apparent.

I think ultimately setting permissions will need to be a post-install procedure that overrides what the build does, so in these functions, we should just preserve what is already there rather than trying to get the permissions "right".

Reply:
I was also confused when I saw this. At first I thought `copy_mode` would overwrite everything that `set_install_permissions` did, but then I realized that it only adds permissions, it doesn't remove them.

If we drop `set_install_permissions`, isn't `copy_mode` essentially identical to `shutil.copymode`? Actually, `shutil.copy` already copies permissions, so we may not need it at all. In that case, `copy`/`copy_tree` and `install`/`install_tree` would be identical aside from the debug message, which I would like to keep distinct.

Thoughts?

Reply:
The more I think about this, the more I'm hesitant to change the current permissions logic. It has worked pretty well for us so far. If a package needs to be manually installed, there is no guarantee that it will have world permissions. The current logic ensures that all users will be able to use the installed software, but without that, we have no guarantee.

If @becker33's work will make this obsolete, then so be it, but until then, we should keep the current logic in place.

Reply:
Ok, you're right.  Let's keep it like it is for now and wait for a permissions revamp.  Might be good to change the name of `copy_mode` to something that describes what it does better (`update_mode`? `expand_mode`?).

We really need a configuration option to set the baseline permissions at a coarse granularity (currently defined by `set_install_permissions`)  and `copy_mode` should not update beyond that baseline.  So, if I have a group-private, group-writable spack instance, then copy_mode should *not* add world-readable or world-writable bits on top of that.  We can think about that later, I guess.